tests/unit/module_utils/test_resource.py

```diff
@@ -0,0 +1,241 @@
+# Copyright: (c) 2023, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from unittest.mock import MagicMock, patch, call
+import pytest
+from typing import Dict, List
+
+from ansible_collections.cloud.pravic.plugins.module_utils.resource import (
+    get_value,
+    replace_reference,
+    REREG,
+    resolve_refs,
+    CloudClient,
+    ResourceExceptionError,
+)
+
+PATCH_BASE_PATH = "ansible_collections.cloud.pravic.plugins.module_utils.resource."
+
+
+@pytest.mark.parametrize(
+    "data,path,expected",
+    [
+        ({"a": "value_a"}, ["a"], "value_a"),
+        (
+            {"parent_a": {"child_1": "value_a_1", "child_2": "value_a_2"}},
+            ["parent_a", "child_1"],
+            "value_a_1"
+        ),
+        (
+            {"parent_a": {"child_1": "value_a_1", "child_2": "value_a_2"}},
+            ["parent_a", "child_1", "another_child"],
+            None
+        ),
+    ]
+)
+def test_get_value(data, path, expected):
+    if expected:
+        assert expected == get_value(data, path)
+    else:
+        with pytest.raises(KeyError):
+            get_value(data, path)
+
+
+@pytest.mark.parametrize("data", ["a.b.c", "a.b"])
+@patch(PATCH_BASE_PATH + "get_value")
+def test_replace_reference(m_get_value, data):
+    match = REREG.match(f"resource:{data}")
+    context = MagicMock()
+    value = MagicMock()
+    m_get_value.return_value = value
+    assert replace_reference(context, match) == value
+    assert m_get_value.called_once_with(call(context, data.split(".")))
+
+
+@pytest.mark.parametrize(
+    "node,expected",
+    [
+        ({"a": "resource:s1.a"}, {"a": "s1_value_a"}),
+        ({"a": "resource:s1.a", "b": "resource:s2.b"}, {'a': 's1_value_a', 'b': 's2_value_b'}),
+        ({"c": ["resource:s1.a", "resource:s2.b"]}, {'c': ['s1_value_a', 's2_value_b']}),
+        ({"a": "resource:s1.c"}, None),
+    ]
+)
+@patch(PATCH_BASE_PATH + "replace_reference")
+def test_resolve_refs(m_replace_reference, node, expected):
+
+    context = {
+        "s1": {"a": "s1_value_a", "b": "s1_value_b"},
+        "s2": {"a": "s2_value_a", "b": "s2_value_b"},
+    }
+
+    def replace_ref(ctxt, match):
+        tmp = ctxt
+        for k in match.group(1).split("."):
+            tmp = tmp[k]
+        return tmp
+
+    m_replace_reference.side_effect = replace_ref
+    check_mode = False
+
+    if expected is None:
+        with pytest.raises(KeyError):
+            resolve_refs(node, context, check_mode=check_mode)
+    else:
+        result = resolve_refs(node=node, context=context, check_mode=check_mode)
+        print(f"Result: {result} - Expected: {expected}")
+        assert result == expected
+
+
+@patch(PATCH_BASE_PATH + "replace_reference")
+def test_resolve_refs_with_check_mode(m_replace_reference):
+
+    context = {}
+
+    def raise_key_error():
+        return KeyError("missing key from dictionnary")
+
+    m_replace_reference.side_effect = raise_key_error()
```

Comment:
Can't you just do `m_replace_reference.side_effect = KeyError('some key')`?

Reply:
done